fix: skip empty Kafka partitions when calculating pending count

KeranYang opened the pull request:
closes #642 

### The problem
When source Kafka topic has multiple partitions and user is sending very few messages, there can be cases when some of the partitions are empty.

For partitions that are empty (let's call those idle partitions), `sarama.ClusterAdmin` `ListConsumerGroupOffsets` API returns us block offset of `-1` to indicate there is no offset associated with that specific partition. Meanwhile when we call `r.saramaClient.GetOffset(r.topic, partition, sarama.OffsetNewest)` to get the newest available offset for idle partitions, it returns us 0. Eventually when we calculate total pending messages using `totalPending += partitionOffset - block.Offset`, for idle partitions, we are always counting (0-(-1)=) 1 for each idle partitions.

Therefore even if we have no input messages at all, if input topic has multiple partitions, the source Kafka vertex will never scale down to 0 pod.

### The fix
When calculating pending count for Kafka, always skip the idle partitions.

### Testing
* Before this change, kafka source vertex doesn't scale down when there is no traffic.
* After this change, I verified it can scale down to 0 pods and scale up when I start sending messages.

<!--

Before you push your changes:

* Run `make pre-push -B` to fix codegen and lint problems.

Then, you MUST:

* Sign-off your commit (otherwise the DCO check will fail).
* Use [a conventional commit message](https://www.conventionalcommits.org/en/v1.0.0/) (otherwise the commit message check will fail).

If you did not do this, reset all your commit and replace them with a single commit:

```
git reset HEAD~1 ;# change 1 to how many commits you made
git commit --signoff -m 'feat: my feat. Fixes #1234'
```

When creating your PR: 

* Make sure that "Fixes #" or "Closes #" is in both the PR title (for release notes) and description (to automatically link and close the issue).
* Say how you tested your changes. If you changed the UI, attach screenshots.
* Set your PR as a draft initially.
* Your PR needs to pass the required checks before it can be approved. 
* Once required tests have passed, mark your PR "Ready for review".

If changes were requested, once you've made them, you MUST dismiss the review to get it reviewed again.

-->

whynowy commented:
Nice!

whynowy commented:
> closes #642
> 
> ### The problem
> When source Kafka topic has multiple partitions and next vertex has one single pod, only one of the partitions is actively being consumed.

- Do you mean some of the partitions has no data?

> 
> For partitions that are not consumed (let's call those idle partitions), `sarama.ClusterAdmin` `ListConsumerGroupOffsets` API returns us block offset of `-1` to indicate there is no offset associated with that specific partition. Meanwhile when we call `r.saramaClient.GetOffset(r.topic, partition, sarama.OffsetNewest)` to get the newest available offset for idle partitions, it returns us 0. Eventually when we calculate total pending messages using `totalPending += partitionOffset - block.Offset`, for idle partitions, we are always counting (0-(-1)=) 1 for each idle partitions.
> 
> Therefore even if we have no input messages at all, if input topic has multiple partitions, the source Kafka vertex will never scale down to 0 pod.
> 
> ### The fix
> When calculating pending count for Kafka, always skip the idle partitions.
> 
> ### Testing
> * Before this change, kafka source vertex doesn't scale down when there is no traffic.
> * After this change, I verified it can scale down to 0 pods and scale up when I start sending messages.

vigith commented:
Nice work! 

> When source Kafka topic has multiple partitions and next vertex has one single pod, only one of the partitions is actively being consumed.

Are you saying if we have 2 partitions with one consumer pod in the source vertex, then there will be a partition whose data will never be consumed?

KeranYang commented:
> Do you mean some of the partitions have no data?
@whynowy Yes. I updated the PR description to reflect the fact.

> Are you saying if we have 2 partitions with one consumer pod in the source vertex, then there will be a partition whose data will never be consumed?
@vigith Sorry I put some mis-leading descriptions there. I have updated the PR description. The real case is that there can be partitions with no data.

whynowy commented on pkg/sources/kafka/reader.go:
```diff
@@ -254,11 +254,16 @@ func (r *KafkaSource) Pending(ctx context.Context) (int64, error) {
 	for _, partition := range partitions {
+		block := rep.GetBlock(r.topic, partition)
+		if block.Offset == -1 {
+			// This partition has not been processed yet.
```
This comment is still misleading, update it together with some other PRs.

vigith replied:
I thought it was valid, thinking what the comment meant was this partition had never been processed; hence it is -1. If it ever read something, it wouldn't be -1.

@KeranYang is this understanding correct?

KeranYang replied:
@vigith Yes your understanding is correct. I think the word `process` can be misleading as people can think that `not been processed`  means the partition has data but not being consumed yet. I can add more details and explicitly mention the partition has no data.
